clang/include/clang/Sema/Sema.h

```diff
@@ -1825,6 +1825,11 @@ class Sema final : public SemaBase {
   /// Set of no-builtin functions listed by \#pragma function.
   llvm::SmallSetVector<StringRef, 4> MSFunctionNoBuiltins;
 
+  /// Map of BuiltinIDs to source locations that have #pragma intrinsic calls
+  /// that refer to them.
+  llvm::DenseMap<unsigned, llvm::SmallSetVector<SourceLocation, 4>>
```

Comment:
DenseMap of SmallVector values is not an efficient data structure choice, because the DenseMap is always at least 25% empty, and each SmallVector is likely to contain exactly 1 element.

I think we need to serialize these pragmas through the AST. Is it possible to do this by creating an implicit declaration of the builtin at the point of the pragma? You would test for this by making your system header into a module or PCH file in another test case.

Reply:
Sorry, I intended to write a comment explaining this commit but I got distracted and you reviewed it too fast :)

I will say that I tried creating the declaration just by calling `LazilyCreateBuiltin` in the pragma handler, but that caused problems when there is a declaration in the header because then we get a ambigous call error. The reason that doesn't happen today with no changes is because we only call `LazilyCreateBuiltin` when the lookup fails, ex when there is no header providing a declaration. 

Let me try your idea, thanks for the feedback.

Reply:
I tried compiling the system header into a PCH and [dumping](https://gist.github.com/sarnex/2ae0c9c543a004622734a6cb1bdead0a) the AST with `llvm-bcanalyzer` and I didn't see any implicit declarations, not sure if I was checking for the right thing.



Either way, I tried another solution where we declare the builtin if it wasn't already delcared. Fixes the ambiguous call errors I saw when trying to universally declare it and fixes the original issue. Please take a look when you have a sec, thanks.